moshi/src/main/java/com/squareup/moshi/JsonWriter.kt

```diff
@@ -0,0 +1,543 @@
+  @Throws(IOException::class)
+  public abstract fun name(name: String): JsonWriter
+
+  /**
+   * Encodes `value`.
+   *
+   * @param value the literal string value, or null to encode a null literal.
+   * @return this writer.
+   */
+  @Throws(IOException::class)
+  public abstract fun value(value: String?): JsonWriter
+
+  /**
+   * Encodes `null`.
+   *
+   * @return this writer.
+   */
+  @Throws(IOException::class)
+  public abstract fun nullValue(): JsonWriter
+
+  /**
+   * Encodes `value`.
+   *
+   * @return this writer.
+   */
+  @Throws(IOException::class)
+  public abstract fun value(value: Boolean): JsonWriter
+
+  /**
+   * Encodes `value`.
+   *
+   * @return this writer.
+   */
+  @Throws(IOException::class)
+  public abstract fun value(value: Boolean?): JsonWriter
+
+  /**
+   * Encodes `value`.
+   *
+   * @param value a finite value. May not be [NaNs][Double.isNaN] or [infinities][Double.isInfinite].
+   * @return this writer.
+   */
+  @Throws(IOException::class)
+  public abstract fun value(value: Double): JsonWriter
+
+  /**
+   * Encodes `value`.
+   *
+   * @return this writer.
+   */
+  @Throws(IOException::class)
+  public abstract fun value(value: Long): JsonWriter
+
+  /**
+   * Encodes `value`.
+   *
+   * @param value a finite value. May not be [NaNs][Double.isNaN] or [infinities][Double.isInfinite].
+   * @return this writer.
+   */
+  @Throws(IOException::class)
+  public abstract fun value(value: Number?): JsonWriter
+
+  /**
+   * Writes `source` directly without encoding its contents. Equivalent to
+   * ```java
+   * try (BufferedSink sink = writer.valueSink()) {
+   *   source.readAll(sink):
+   * }
+   * ```
+   *
+   * @see valueSink
+   */
+  @Throws(IOException::class)
+  public fun value(source: BufferedSource): JsonWriter {
+    check(!promoteValueToName) { "BufferedSource cannot be used as a map key in JSON at path $path" }
+    valueSink().use(source::readAll)
+    return this
+  }
+
+  /**
+   * Returns a [BufferedSink] into which arbitrary data can be written without any additional
+   * encoding. You **must** call [BufferedSink.close] before interacting with this `JsonWriter` instance again.
+   *
+   * Since no validation is performed, options like [serializeNulls] and other writer
+   * configurations are not respected.
+   */
+  @CheckReturnValue
+  @Throws(IOException::class)
+  public abstract fun valueSink(): BufferedSink
+
+  /**
+   * Encodes the value which may be a string, number, boolean, null, map, or list.
+   *
+   * @return this writer.
+   * @see JsonReader.readJsonValue
+   */
+  @Throws(IOException::class)
+  public fun jsonValue(value: Any?): JsonWriter {
+    when (value) {
+      is Map<*, *> -> {
+        beginObject()
+        for ((key, value1) in value) {
+          require(key is String) {
+            if (key == null) {
+              "Map keys must be non-null"
+            } else {
+              "Map keys must be of type String: ${key.javaClass.name}"
+            }
```

Comment:
nit -- i feel like this'd be more idiomatic as 2 `require`s? 
```suggestion
		 requireNotNull(key) { "Map keys must be not-null" } 
		 require(key is String) { "Map keys must be of type String: ${key.javaClass.name}" }
```

Reply:
dig, though github won't let me add suggestion from here. Let me push

Reply:
16de888